feat: read-only geolocation (GDE-86)

barredterra opened the pull request:
Continues work from #16513

Depends on #21055

- Render the map into the control's `display_area`
- Hide draw controls if the map is read-only
- Remove the useless refresh button
- Separate `get_leaflet_controls` and `point_to_layer` into separate methods that can be more easily overwritten with a custom script

## Before

### Read-Write

![before-rw](https://user-images.githubusercontent.com/14891507/162623951-d3ca7388-f106-4fb4-ae49-59a1bfa1d567.png)

### Read only

![before-r](https://user-images.githubusercontent.com/14891507/162623958-e97d8e48-c102-4d3f-96fd-786769578cf4.png)

## After
### Read-Write

![after-rw](https://user-images.githubusercontent.com/14891507/162623735-b3f14b2b-92dd-4287-84be-57532afaae62.png)

### Read only

![after-r](https://user-images.githubusercontent.com/14891507/162623737-6e29fdab-e767-4728-ab26-e2f2332e68b9.png)

## Todo

- [x] map not displayed after hard-refresh (<kbd>CMD</kbd> + <kbd>SHIFT</kbd> + <kbd>R</kbd>)

    - Map is displayed after `cur_frm.reload_doc()`. Maybe this doesn't get triggered on hard refresh?
    - No problems in version 13
- [x] read-only map doesn't load on empty data (#21088)
- [x] `this.map.invalidateSize()` probably needs to get called at a different time. (https://github.com/frappe/frappe/pull/16561#issuecomment-1570225211)


## Scenarios to test

Does the map render nicely in all of the following scenarios?

- Map in collapsible and normal section
- Refresh, hard-refresh, or page through records
- Editable and read-only map
- With and without data/drawings

> no-docs

Internal reference: LAN-756

barredterra committed:
feat: read-only geolocation (GDE-86)

- render map into display_area
- hide draw controls if read-only
- remove useless refresh_button

barredterra committed:
Merge branch 'develop' into map-read-only

shariquerik commented:
@barredterra Are you working on the Todo part you mentioned?

barredterra commented:
@shariquerik I'm currently out of ideas / don't know where to start searching 😅. Do you have any ideas?

shariquerik commented:
> @shariquerik I'm currently out of ideas / don't know where to start searching 😅. Do you have any ideas?

I will have to look into it.
Btw hide draw controls if read-only is not working.

barredterra commented:
> Btw hide draw controls if read-only is not working.

Are you sure you're loading the most recent code? Tested this several times and using it in production v13 already.

barredterra committed:
Merge branch 'develop' into map-read-only

barredterra committed:
Merge branch 'develop' into map-read-only

barredterra committed:
Merge branch 'develop' into map-read-only

hrwX commented on frappe/public/js/frappe/form/controls/geolocation.js:
```diff
@@ -6,69 +6,69 @@ frappe.ui.form.ControlGeolocation = class ControlGeolocation extends frappe.ui.f
-							return L.marker(latlng);
-						}
+
+		var data_layers = new L.FeatureGroup()
```
```suggestion
		let data_layers = new L.FeatureGroup()
```

hrwX commented:
@barredterra Am I missing something here ? When I make this field read_only, the field disappears

https://user-images.githubusercontent.com/7310479/167878676-6d7a80da-6740-42a3-ab37-fd5528ed3465.mp4

hrwX commented:
@barredterra the `refresh_input` in `base_control.js` is not triggered => `set_input` in `data.js` is not triggered => `set_disp_area` in `data.js` is not triggered and hence we cannot see the map on <kbd>CMD</kbd> + <kbd>SHIFT</kbd> + <kbd>R</kbd> press

barredterra committed:
Merge branch 'develop' into map-read-only

surajshetty3416 commented:
@barredterra this still has that refresh issue... are you working on this? If not, please enable "Allow edits from maintainers" for this so that required changes can be pushed.

barredterra commented:
@surajshetty3416 "Allow edits from maintainers" doesn't work for organization-owned repos, but I invited you as a collaborator. Alternatively, anyone can always create a PR against this branch.

barredterra committed:
Merge branch 'develop' into map-read-only

marination committed:
fix: Load map libraries at build time to avoid async issues during geolocation render

- When gelocation control is built, it awaits loading of libraries
- Withi the control everything is await but a layer above where multiple controls are sequentially built this breaks
- Form render goes ahead without waiting for the gelocation control
- The `onload_post_render` in `Location` cannot find the map wrapper as the control is still being built
- Therefore, on a hard load, the control does not show up and appears on soft reload.

barredterra committed:
Merge branch 'develop' into map-read-only

barredterra committed:
refactor: const instead of var, indentation

barredterra committed:
fix: handle read only property

blaggacao commented:
@barredterra Looks like this would need a rebase after https://github.com/frappe/frappe/pull/21055 is merged, now.

As I needed the patch, I've done a rebase [here](https://github.com/frappe/frappe/pull/21144).

On this occasion: Whats the rationale of using `set_disp_area(value)` over `format_for_input(value)`?

barredterra committed:
Merge branch 'develop' into map-read-only

barredterra commented:
> Whats the rationale of using set_disp_area(value) over format_for_input(value)? – @blaggacao 

Long time since i wrote this... 🤔 From the naming, I guess `format_for_input` was not getting called for read-only fields.

barredterra commented:
@marination could you review this PR?

marination commented:
@barredterra Replicated the `this.map.invalidateSize() ` issue. It occurs on geolocation controls that are inside collapsible sections. This is the only pending issue in this PR. Code looks fine and functions fine.

What happens is:
- `invalidateSize()` essentially tries to find the width and height of the map wrapper so that it can make the map occupy the entire wrapper space.
- If the control is in a collapsed section, the wrapper is hidden and therefore `invalidateSize()` gets a width and height of 0 and thus it essentially does nothing.
- When you expand the section, some grey tiles are visible due to this `invalidateSize()` being ineffective

So the "right" time to invoke `invalidateSize()` is when the map control is not hidden anymore i.e. on the `collapse()` event of the section

marination commented on frappe/public/js/frappe/form/controls/geolocation.js:
```diff
@@ -44,52 +46,49 @@ frappe.ui.form.ControlGeolocation = class ControlGeolocation extends frappe.ui.f
+			})
+		);
+		this.add_non_group_layers(data_layers, this.editableLayers);
+		try {
+			this.map.fitBounds(this.editableLayers.getBounds(), {
+				padding: [50, 50],
+			});
+		} catch (err) {
+			// suppress error if layer has a point.
 		}
+		this.editableLayers.addTo(this.map);
 		this.map.invalidateSize();
```
**Suggested Fix the for `invalidateSize` issue explained in https://github.com/frappe/frappe/pull/16561#issuecomment-1570225211** (sorry for the long suggestion, these fixes are dependent on the changes in this PR)

```suggestion
		this.editableLayers.addTo(this.map);
		this.fit_and_recenter_map();
```

Add the following function to the control:
```js
fit_and_recenter_map() {
	// Spread map across the wrapper, recenter and zoom w.r.t bounds
	try {
		this.map.invalidateSize();
		this.map.fitBounds(this.editableLayers.getBounds(), {
			padding: [50, 50],
		});
	} catch (err) {
		// suppress error if layer has a point.
	}
}
```

Now in `section.js` > `collapse()` add this action for geolocation control:
```js
// refresh signature fields
this.fields_list.forEach((f) => {
	if (f.df.fieldtype == "Signature") {
		f.refresh();
	}
	
	// fix
	if (f.df.fieldtype == "Geolocation") {
		f.fit_and_recenter_map();
	}
});
```

barredterra committed:
refactor: call control's `on_section_collapse()`

This way, the logic can stay in the control itself. Each control can
decide what it needs to do on section collapse/expand.

barredterra committed:
fix: fit and recenter map when section is expanded

barredterra committed:
refactor: move pointToLayer into a separate method

For better customizability

barredterra committed:
Merge branch 'develop' into map-read-only

barredterra commented:
@marination thanks a lot for your investigation and the suggested fix. Everything seems to be working fine now. 👍🏼

marination commented:
@barredterra tested all scenarios. LGTM. @surajshetty3416 can you give it a final look and merge ?

surajshetty3416 commented:
LGTM. Merging.
